source/common/router/router.cc

```diff
@@ -151,13 +152,37 @@ FilterUtility::finalTimeout(const RouteEntry& route, Http::HeaderMap& request_he
   }
   timeout.per_try_timeout_ = route.retryPolicy().perTryTimeout();
 
-  Http::HeaderEntry* header_timeout_entry = request_headers.EnvoyUpstreamRequestTimeoutMs();
   uint64_t header_timeout;
-  if (header_timeout_entry) {
-    if (absl::SimpleAtoi(header_timeout_entry->value().getStringView(), &header_timeout)) {
-      timeout.global_timeout_ = std::chrono::milliseconds(header_timeout);
+
+  if (respect_expected_rq_timeout) {
+    // Check if there is timeout set by egress Envoy.
+    // If present, use that value as route timeout and don't override
+    // *x-envoy-expected-rq-timeout-ms* header. At this point *x-envoy-upstream-rq-timeout-ms*
+    // header should have been sanitized by egress Envoy.
+    Http::HeaderEntry* header_expected_timeout_entry =
+        request_headers.EnvoyExpectedRequestTimeoutMs();
+    if (header_expected_timeout_entry) {
+      if (absl::SimpleAtoi(header_expected_timeout_entry->value().getStringView(),
+                           &header_timeout)) {
+        timeout.global_timeout_ = std::chrono::milliseconds(header_timeout);
```

Comment:
nit: can you lift this part out into a helper which takes the header, etc.? It's repeated 3 times in this function and it would help readability.